keps/sig-api-machinery/4988-serve-pagination-from-cache/README.md

```diff
@@ -0,0 +1,401 @@
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Enabling pagination typically improves performance by reducing response sizes.
+However, the kube-apiserver's caching mechanism (watchcache) does not currently
+support pagination. When a client requests a paginated list, the limit parameter
+might be ignored, or the request might bypass the cache and be served directly
+from etcd, leading to performance issues. This document proposes an enhancement
+to the kube-apiserver's caching layer to enable efficient serving of paginated
+lists from the cache, improving performance and predictability.
+
+## Motivation
+
+The kube-apiserver's watchcache is an eventually consistent cache of the cluster
+state. It allows the API server to serve read requests without constantly hitting
+etcd. However, watchcache is not designed for pagination.
+
+Pagination allows clients to request large lists of resources in smaller chunks.
+The client specifies a limit (maximum number of objects per page) and receives
+a continue token in the response. This token is then used in subsequent requests
+to fetch the next page.
+
+To serve paginated requests correctly, the API server needs access to a consistent
+snapshot of the data from which the initial page was generated. Watchcache does
+not store historical data. Therefore, currently, the API server does the following:
+
+- For `resourceVersion=""`: Serve paginated requests directly from etcd, bypassing the cache.
+- For `resourceVersion="0"`: Ignore the limit parameter and return the entire list in a single response.
+
+Both options are problematic:
+
+- Serving from etcd negates the benefits of the cache,
+  leading to increased load on both the API server and etcd.
+- Ignoring the limit parameter defeats the purpose of pagination,
+  potentially returning massive responses that can overwhelm clients.
+
+This situation leads to unexpected behavior for users,
+where enabling pagination can degrade performance instead of improving it.
+
+### Goals
+
+- Improve the performance of paginated list requests to be comparable to non-paginated lists.
+- Make pagination behavior predictable and consistent.
+- Reduce load on etcd caused by paginated requests.
+
+### Non-Goals
+
+- Serve `resourceVersion="N"` request from watch cache
+- Support indices when paginating.
+- Eliminate all paginated list requests to etcd.
+
+## Proposal
```

Comment:
I just very briefly skimmed through it and I don't see alternatives.
The main alternative that I have is that:

1) we do btree.Clone() on every watch event (and store it until TTL)  - i.e. independently if there was a request about it or not
2) we can now serve every request (e.g. different kube-apiserver risk goes away) up to TTL from cache from any kube-apiserver (we just use largest RV not greater then the provided one)

Unless the overhead of it is substantial (which I don't know) - that would be my preferred option. But I think we need some experiment to estimate the overhead.

Reply:
Right, I have implemented something similar for serving exact revision and was planning to as next step, however you are right that we should have a better understanding of the overhead to know the tradeoffs.

Reply:
FWIW - it might be worth referencing https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/365-paginated-lists#potential-future-extensions because it's mentioning a lot of that and the corresponding PR has some historical context on it

Reply:
I run the scalability tests, the results show 7-18% CPU and 4-6 Memory regression and regression in 99%ile of single object operations. Current scalability tests don't utilize pagination or exact requests at all, so this should be the closes to measuring the overhead. However it's not great at showing us tradeoff, before deciding next step we would need to update scalability tests to be more realistic depiction of workloads using pagination.

Reply:
> FWIW - it might be worth referencing https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/365-paginated-lists#potential-future-extensions because it's mentioning a lot of that and the corresponding PR has some historical context on it

Makes sense

Reply:
Repeated the test looks like the overhead is small. Small enough that it might be acceptable option.